File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int currmax = nums[0];
        int currmin = nums[0];
        int compare = nums[0];

        for (int i = 1; i < nums.length; i++) {
        int temp = currmax;
        currmax = Math.max(nums[i], Math.max(currmax*nums[i], currmin*nums[i]));
        currmin = Math.min(nums[i], Math.min(temp*nums[i], currmin*nums[i]));
        compare = Math.max(compare, currmax);
        }

        return compare; 
    }
}

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        curmx=curmn=mx=mn=nums[0]
        for i in range(1,len(nums)):
            if nums[i]>0:
                curmx,curmn=max(curmx*nums[i],nums[i]),min(curmn*nums[i],nums[i])
            else:
                curmx,curmn=max(curmn*nums[i],nums[i]),min(curmx*nums[i],nums[i]) 
            mx=max(mx,curmx)
            mn=min(mn,curmn)
        return mx